tests/web_server/test_core_endpoints.py

```diff
@@ -140,3 +141,53 @@ async def test_command_overwrite_failure(aiohttp_client, authorized_cookies):
     assert resp.status == HTTPStatus.OK
     config_dict = await resp.json()
     assert config_dict.get('requirements', dict()).get('go', dict()).get('command') == 'go version'
+
+
+async def test_custom_rejecting_login_handler(aiohttp_client):
+    class RejectAllLoginHandler(LoginHandlerInterface):
+        def __init__(self, services):
+            self.auth_svc = services.get('auth_svc')
+            self._name = 'Reject All Logins'
+
+        """ LoginHandlerInterface implementation """
+
+        async def handle_login(self, request, **kwargs):
+            # Always reject login and return 401
+            raise web.HTTPUnauthorized(text='Automatic rejection')
+
+        async def handle_login_redirect(self, request, **kwargs):
+            # Always reject and return 401
+            raise web.HTTPUnauthorized(text='Automatic rejection')
+
+    BaseService.get_service('auth_svc')._login_handler = RejectAllLoginHandler(BaseService.get_services())
+    resp = await aiohttp_client.post('/enter', allow_redirects=False, data=dict(username='admin', password='admin'))
+    assert resp.status == HTTPStatus.UNAUTHORIZED
+    assert await resp.text() == 'Automatic rejection'
+
+    resp = await aiohttp_client.get('/', allow_redirects=False)
+    assert resp.status == HTTPStatus.UNAUTHORIZED
+    assert await resp.text() == 'Automatic rejection'
+
+
+async def test_custom_accepting_login_handler(aiohttp_client):
+    class AcceptAllLoginHandler(LoginHandlerInterface):
+        def __init__(self, services):
+            self.auth_svc = services.get('auth_svc')
+            self._name = 'Accept All Logins'
+
+        """ LoginHandlerInterface implementation """
+
+        async def handle_login(self, request, **kwargs):
+            # Always accept login
+            data = await request.post()
+            username = data.get('username', 'default username')
+            await self.auth_svc.provide_verified_login_response(request, username)
+
+        async def handle_login_redirect(self, request, **kwargs):
+            await self.handle_login(request, **kwargs)
+
+    BaseService.get_service('auth_svc')._login_handler = AcceptAllLoginHandler(BaseService.get_services())
```

Comment:
Seeing the setting of `_login_handler` makes me think this should be supported via a specific interface (e.g., set_login_handler(...)`. Perhaps the auth service shouldn't necessarily deal with the configuration itself and instead support constructor injection of handlers. There could potentially be another factory (e.g., a classmethod or something) that knows how to deal with the configuration, load/import things and then correctly instantiate/parameterize the auth service. Just spitballin' here.

Reply:
I'll have a public method `set_login_handlers(services)` that'll get called during `run_tasks` in `server.py`

Reply:
`set_login_handlers` will either get the handler from the config file or set a handler object passed in as a parameter